src/routes/relay/services/relay-limit.service.ts

```diff
@@ -26,47 +30,65 @@ export class RelayLimitService {
     return `${chainId}:${ethers.getAddress(address)}`;
   }
 
+  /**
+   * Get the number of cached attempts for an address
+   */
+  private async getCachedAttempts(
+    chainId: string,
+    address: string,
+  ): Promise<number> {
+    const key = this.generateKey(chainId, address);
+    const attempts = await this.cacheService.get(key);
+    return typeof attempts === 'string' ? Number(attempts) : 0;
+  }
+
   /**
    * Get the current relay limit for an address
    */
-  public getRelayLimit(
+  public async getRelayLimit(
     chainId: string,
     address: string,
-  ): {
+  ): Promise<{
     limit: number;
     remaining: number;
-    expiresAt?: number;
-  } {
-    const key = this.generateKey(chainId, address);
-    const throttlerEntry = this.throttlerStorageService.storage[key] || {
-      totalHits: 0,
-    };
+  }> {
+    const attempts = await this.getCachedAttempts(chainId, address);
 
     return {
       limit: this.limit,
-      remaining: Math.max(0, this.limit - throttlerEntry.totalHits),
-      expiresAt: throttlerEntry.expiresAt,
+      remaining: Math.max(0, this.limit - attempts),
     };
   }
 
   /**
    * Check if addresses can relay
    */
-  public canRelay(chainId: string, addresses: string[]): boolean {
-    return addresses.every((address) => {
-      const limit = this.getRelayLimit(chainId, address);
-      return limit.remaining > 0;
-    });
+  public async canRelay(
+    chainId: string,
+    addresses: Array<string>,
+  ): Promise<boolean> {
+    const attempts = await Promise.all(
+      addresses.map((address) => this.getCachedAttempts(chainId, address)),
+    );
+    return attempts.every((attempts) => attempts < this.limit);
   }
 
   /**
    * Increment the number of relays for addresses
    */
-  public async increment(chainId: string, addresses: string[]): Promise<void> {
+  public async increment(
+    chainId: string,
+    addresses: Array<string>,
+  ): Promise<void> {
+    const attempts = await Promise.all(
+      addresses.map((address) => this.getCachedAttempts(chainId, address)),
+    );
+
     await Promise.all(
-      addresses.map((address) => {
+      addresses.map((address, i) => {
         const key = this.generateKey(chainId, address);
-        return this.throttlerStorageService.increment(key, this.ttl);
+        const incremented = attempts[i] + 1;
+        return this.cacheService.set(key, incremented.toString(), this.ttl);
```

Comment:
I think we can merge the two steps above. What do you think about the following?

```typescript
await Promise.all(
      addresses.map(async (address) => {
        const key: string = this.generateKey(chainId, address);
        const currentAttempts: number = await this.getCachedAttempts(
          chainId,
          address,
        );
        const incremented: number = currentAttempts + 1;
        return this.cacheService.set(key, incremented.toString(), this.ttl);
      }),
    );
```

Another thing to consider is in the same way that we have a `getCachedAttempts` returning a number we can probably have a `setCachedAttempts` which takes a number. With that, we "normalize" the read/write access to the cache and hide the `string` representation behind these two functions.

One thing to consider though is that the attempts might get partially updated – if one fails this promise is rejected. What do you think about using `Promise.allSettled()` instead?

Reply:
Great ideas! I've combined them all as follows:

- There is now a `setCachedAttempts` function.
- `increment` leverages this, combining both the getting and setting inside one `map`.
- All of which is wrapped in a `Promise.allSettled`.